### PR TITLE
server: cannot deploy/start vm if service offering has multiple tags

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes an issue in #4399 

Steps to reproduce the issue
(1) create service offering with multiple host tags (2 or more)
(2) create a vm with the service offering

expected result
vm is started or 
get error message "No suitable host found for follow compute offering tags: tag1,tag2"

actual result:
Got exception
```
com.cloud.utils.exception.CloudRuntimeException: DB Exception on: select distinct cluster_id from host join host_tags on host.id = host_tags.host_id and host_tags.tag = ?
```

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by nvazquez — APPROVED** [review]
LGTM - Manually verified

**Comment by nvazquez** [comment]
Ping for review @sureshanaparti @davidjumani @rhtyd

**Review by rhtyd — APPROVED** [review]

File: engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java [hunk 1]
```diff
@@ -85,7 +85,7 @@
                                                              + "GROUP BY host_id "
                                                              + "HAVING tag_count = %s ";
     private static final String SEPARATOR = ",";
-    private static final String LIST_CLUSTERID_FOR_HOST_TAG = "select distinct cluster_id from host join host_tags on host.id = host_tags.host_id and host_tags.tag = ?";
+    private static final String LIST_CLUSTERID_FOR_HOST_TAG = "select distinct cluster_id from host join ( %s ) AS selected_hosts ON host.id = selected_hosts.host_id";
```
Comment (rhtyd) [inline engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java:88]: Should `%s` be replaced with `? `

Comment (sureshanaparti) [reply engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java:88]: @rhtyd It seems, all the tags are part of the sub-query stmt "_LIST_HOST_IDS_BY_COMPUTETAGS_" and the returned hosts from that sub-query were considered from further filtering.

**Review by sureshanaparti — APPROVED** [review]
LGTM

File: engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java (hunk 1 above)
Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java:88]: > Should `%s` be replaced with `? `

@rhtyd 
see suresh's comment above.

https://github.com/apache/cloudstack/blob/fccd02b6ced1ebe88a7bfd606297f339e119d5c5/engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java#L1208-L1213

an example of final sql is (if there are 3 tags)
```
select distinct cluster_id from host join 
( SELECT filtered.host_id, COUNT(filtered.tag) AS tag_count FROM (SELECT host_id, tag FROM host_tags GROUP BY host_id,tag) AS filtered WHERE tag IN(?,?,?) GROUP BY host_id HAVING tag_count = 3  ) AS selected_hosts 
ON host.id = selected_hosts.host_id;
```

**Pull request merged**